ra/dsl.py

```diff
@@ -0,0 +1,419 @@
+
+    def put(self, test_fn=None, **req_params):
+        "Decorator for defining PUT method tests, like ``ResourceScope.get``"
+        return self.method('put', test_fn=test_fn, **req_params)
+
+    def patch(self, test_fn=None, **req_params):
+        "Decorator for defining PATCH method tests, like ``ResourceScope.get``"
+        return self.method('patch', test_fn=test_fn, **req_params)
+
+    def delete(self, test_fn=None, **req_params):
+        "Decorator for defining DELETE method tests, like ``ResourceScope.get``"
+        return self.method('delete', test_fn=test_fn, **req_params)
+
+    def head(self, test_fn=None, **req_params):
+        "Decorator for defining HEAD method tests, like ``ResourceScope.get``"
+        return self.method('head', test_fn=test_fn, **req_params)
+
+    def options(self, test_fn=None, **req_params):
+        "Decorator for defining OPTIONS method tests, like ``ResourceScope.get``"
+        return self.method('options', test_fn=test_fn, **req_params)
+
+
+class TestSuite(object):
+    """Used internally to log when tests are added for a resource.
+    """
+    def __init__(self):
+        self.tests = []
+
+    def add_test(self, test, resource_node):
+        self.tests.append((test, resource_node))
+
+    def test_exists(self, method, path):
+        for test, resource_node in self.tests:
+            if resource_node.method == method and resource_node.path == path:
+                return True
+        return False
+
+
+def _parse_raml(raml_path_or_string):
+    "Returns the RAML file path (or None if arg is a string) and parsed RAML."
+    raml_path = None
+    if not raml.is_raml(raml_path_or_string):
+        raml_path = raml_path_or_string
+
+    parsed_raml = raml.parse(raml_path_or_string)
+
+    return raml_path, parsed_raml
+
+
+def make_request_class(app, base=None):
+    """Create a callable, app-bound request class from a base request class.
+
+    Request objects built from this class are passed to test functions.
+    The request object is callable with this signature:
+
+        req(validate=True, **req_params)
+
+    When called, a request is made against app, passing the object as
+    the request to send, applying any additional request parameters passed.
+    It also takes a ``validate`` keyword to determine if RAML validation
+    assertions should be automatically made on the response before
+    returning it (default is True).
+
+    The request object expects to be assigned a ``raml`` attribute with
+    the ra.raml.ResourceNode to validate against as the value.
+
+    :param app:         the app we want to make requests to, generally an
+                        instance of ``webtest.TestApp`` but can be anything
+                        that responds to request() taking a webob-like request
+                        object as a first positional argument, and accepts
+                        request parameters as keyword args.
+    :param base:        the base request class
```

Comment:
Good catch. Fixed and also simplified to just use webtest's request/response (better methods on response for testing).